File: backend/app/render/zones.py
"""Grille fixe du dépliant : le document n'est jamais pensé comme une
succession de Paragraph empilés, mais comme une page imprimée composée de
zones à coordonnées fixes (comme dans InDesign/Publisher/Scribus). Chaque
Zone a une position, une largeur et une hauteur invariables — le contenu
(chants) est ensuite injecté dans ces zones, jamais l'inverse."""
from dataclasses import dataclass

from reportlab.lib.pagesizes import A4, landscape
from reportlab.lib.units import mm

PAGE_SIZE = landscape(A4)
PAGE_W, PAGE_H = PAGE_SIZE

MARGE = 5 * mm
ENTRE_COLONNES = 6 * mm
EPAISSEUR_BORDURE = 0.4

HAUTEUR_ENTETE = 46 * mm
"""Bloc fixe en haut du demi-page droite de la page 1 (logos, titre,
chorale, date, lectures) — jamais déplacé."""

HAUTEUR_BANNIERE = 45 * mm
"""Bande fixe en bas du demi-page gauche de la page 1 (annonce, bannière
décorative, coordonnées) — toujours présente, jamais utilisée pour les chants."""

X0 = MARGE
Y0 = MARGE
LARGEUR_UTILE = PAGE_W - 2 * MARGE
HAUTEUR_UTILE = PAGE_H - 2 * MARGE

LARGEUR_DEMI = (LARGEUR_UTILE - ENTRE_COLONNES) / 2
LARGEUR_COLONNE = (LARGEUR_DEMI - ENTRE_COLONNES) / 2
# Sur la page 2, 4 colonnes égales occupent toute la largeur utile — avec le
# même entre-colonnes, elles tombent exactement à la même largeur que les
# demi-colonnes de la page 1 (G1/G2/D1/D2), assurant une grille identique
# sur les deux pages.
LARGEUR_COLONNE_P2 = (LARGEUR_UTILE - 3 * ENTRE_COLONNES) / 4

X_GAUCHE = X0
X_DROITE = X0 + LARGEUR_DEMI + ENTRE_COLONNES


@dataclass(frozen=True)
class Zone:
    nom: str
    page: int
    x: float
    y: float
    largeur: float
    hauteur: float
    padding: float = 2.0


def _colonne_x(x_base: float, i: int, largeur: float) -> float:
    return x_base + i * (largeur + ENTRE_COLONNES)


def construire_zone_g(banniere_active: bool = True) -> tuple[Zone, Zone]:
    """G1 et G2 : demi-page gauche de la page 1, au-dessus de la bannière.
    Si la bannière n'est pas active, les colonnes descendent jusqu'au bas de la page."""
    y = Y0 + HAUTEUR_BANNIERE if banniere_active else Y0
    hauteur = HAUTEUR_UTILE - HAUTEUR_BANNIERE if banniere_active else HAUTEUR_UTILE
    g1 = Zone("G1", page=1, x=_colonne_x(X_GAUCHE, 0, LARGEUR_COLONNE), y=y, largeur=LARGEUR_COLONNE, hauteur=hauteur)
    g2 = Zone("G2", page=1, x=_colonne_x(X_GAUCHE, 1, LARGEUR_COLONNE), y=y, largeur=LARGEUR_COLONNE, hauteur=hauteur)
    return g1, g2


def construire_zone_d() -> tuple[Zone, Zone]:
    """D1 et D2 : demi-page droite de la page 1, sous l'en-tête fixe."""
    y = Y0
    hauteur = HAUTEUR_UTILE - HAUTEUR_ENTETE
    d1 = Zone("D1", page=1, x=_colonne_x(X_DROITE, 0, LARGEUR_COLONNE), y=y, largeur=LARGEUR_COLONNE, hauteur=hauteur)
    d2 = Zone("D2", page=1, x=_colonne_x(X_DROITE, 1, LARGEUR_COLONNE), y=y, largeur=LARGEUR_COLONNE, hauteur=hauteur)
    return d1, d2


def construire_zones_page2() -> list[Zone]:
    """4 colonnes strictement identiques occupant toute la page 2."""
    return [
        Zone(f"C{i + 1}", page=2, x=_colonne_x(X0, i, LARGEUR_COLONNE_P2), y=Y0,
             largeur=LARGEUR_COLONNE_P2, hauteur=HAUTEUR_UTILE)
        for i in range(4)
    ]


@dataclass(frozen=True)
class Grille:
    flow_order: list[Zone]
    """Ordre exact de remplissage des chants : D1 -> D2 -> page2 (4 colonnes)
    -> G1 -> G2 (G2 exclue si la Prière pour le Burkina Faso est active)."""
    toutes: dict[str, Zone]
    """Toutes les zones par nom, y compris celles hors du flux (ex: G2 quand
    la Prière l'occupe) — pour que le rendu puisse toujours les localiser."""


def construire_grille(priere_active: bool, one_page_mode: bool = False, banniere_active: bool = True) -> Grille:
    """L'ordre de remplissage ne suit jamais l'ordre physique des pages : le
    moteur remplit des zones, pas des pages. D1/D2 (page 1, sous l'en-tête)
    sont remplies en premier, puis les 4 colonnes de la page 2, et seulement
    en dernier G1/G2 (page 1, au-dessus de la bannière) — jamais l'inverse,
    jamais une autre séquence.
    Si one_page_mode est True, le flux est restreint à la moitié droite de la
    première page (D1/D2) et la partie gauche de la seconde page (C1/C2)."""
    d1, d2 = construire_zone_d()
    g1, g2 = construire_zone_g(banniere_active)

    if one_page_mode:
        # En mode 1 page, C1/C2 sont dessinés sur la moitié gauche de la page 1
        y_c = Y0 + HAUTEUR_BANNIERE if banniere_active else Y0
        h_c = HAUTEUR_UTILE - HAUTEUR_BANNIERE if banniere_active else HAUTEUR_UTILE
        c1 = Zone("C1", page=1, x=_colonne_x(X0, 0, LARGEUR_COLONNE_P2), y=y_c, largeur=LARGEUR_COLONNE_P2, hauteur=h_c)
        c2 = Zone("C2", page=1, x=_colonne_x(X0, 1, LARGEUR_COLONNE_P2), y=y_c, largeur=LARGEUR_COLONNE_P2, hauteur=h_c)
        # c3 and c4 are unused in 1-page mode but initialized to avoid KeyError in toutes
        c3 = Zone("C3", page=1, x=_colonne_x(X0, 2, LARGEUR_COLONNE_P2), y=Y0, largeur=LARGEUR_COLONNE_P2, hauteur=HAUTEUR_UTILE)
        c4 = Zone("C4", page=1, x=_colonne_x(X0, 3, LARGEUR_COLONNE_P2), y=Y0, largeur=LARGEUR_COLONNE_P2, hauteur=HAUTEUR_UTILE)
        zones_p2 = [c1, c2, c3, c4]
    else:
        zones_p2 = construire_zones_page2()

    toutes = {z.nom: z for z in [d1, d2, g1, g2, *zones_p2]}

    if one_page_mode:
        # En mode 1 page, si la prière est active, elle occupe C2. Donc C2 est exclu du flux des chants.
        flow = [d1, d2, c1]
        if not priere_active:
            flow.append(c2)
    else:
        flow = [d1, d2, *zones_p2, g1]
        if not priere_active:
            flow.append(g2)
    return Grille(flow_order=flow, toutes=toutes)

File: mobile/src/render/zones.ts
// Grille fixe du dépliant, portée de backend/app/render/zones.py -- le
// document n'est jamais pensé comme un flux de texte, mais comme une page
// imprimée composée de zones à coordonnées fixes (comme InDesign/Scribus) :
// le contenu (chants) est injecté DANS ces zones, jamais l'inverse.
//
// Coordonnées en mm directement (CSS supporte l'unité `mm` nativement,
// contrairement à ReportLab qui travaille en points) -- mêmes valeurs mm que
// le fichier Python, sans conversion. `y` suit la convention ReportLab
// (origine en bas de page, y croît vers le haut) ; la conversion vers le
// repère CSS (origine en haut) se fait au moment de l'assemblage HTML, voir
// pdfAssembler.ts, jamais ici.

export const PAGE_W_MM = 297; // A4 paysage
export const PAGE_H_MM = 210;

export const MARGE_MM = 5;
export const ENTRE_COLONNES_MM = 6;

export const HAUTEUR_ENTETE_MM = 46;
export const HAUTEUR_BANNIERE_MM = 45;

export const X0_MM = MARGE_MM;
export const Y0_MM = MARGE_MM;
export const LARGEUR_UTILE_MM = PAGE_W_MM - 2 * MARGE_MM;
export const HAUTEUR_UTILE_MM = PAGE_H_MM - 2 * MARGE_MM;

export const LARGEUR_DEMI_MM = (LARGEUR_UTILE_MM - ENTRE_COLONNES_MM) / 2;
export const LARGEUR_COLONNE_MM = (LARGEUR_DEMI_MM - ENTRE_COLONNES_MM) / 2;
// Sur la page 2, 4 colonnes égales occupent toute la largeur utile -- avec le
// même entre-colonnes, elles tombent exactement à la même largeur que les
// demi-colonnes de la page 1, assurant une grille identique sur les 2 pages.
export const LARGEUR_COLONNE_P2_MM = (LARGEUR_UTILE_MM - 3 * ENTRE_COLONNES_MM) / 4;

export const X_GAUCHE_MM = X0_MM;
export const X_DROITE_MM = X0_MM + LARGEUR_DEMI_MM + ENTRE_COLONNES_MM;

// Retrait de sécurité entre le texte et chaque bordure : 0,5 pt.
const PADDING_MM = 0.5 / 2.8346;

export interface Zone {
  nom: string;
  page: 1 | 2;
  x: number;
  y: number;
  largeur: number;
  hauteur: number;
  padding: number;
}

function colonneX(xBase: number, i: number, largeur: number): number {
  return xBase + i * (largeur + ENTRE_COLONNES_MM);
}

function zone(nom: string, page: 1 | 2, x: number, y: number, largeur: number, hauteur: number): Zone {
  return { nom, page, x, y, largeur, hauteur, padding: PADDING_MM };
}

/** G1/G2 : demi-page gauche de la page 1, au-dessus de la bannière (ou
 * jusqu'en bas si elle est désactivée). */
export function construireZoneG(banniereActive = true): [Zone, Zone] {
  const y = banniereActive ? Y0_MM + HAUTEUR_BANNIERE_MM : Y0_MM;
  const hauteur = banniereActive ? HAUTEUR_UTILE_MM - HAUTEUR_BANNIERE_MM : HAUTEUR_UTILE_MM;
  return [
    zone("G1", 1, colonneX(X_GAUCHE_MM, 0, LARGEUR_COLONNE_MM), y, LARGEUR_COLONNE_MM, hauteur),
    zone("G2", 1, colonneX(X_GAUCHE_MM, 1, LARGEUR_COLONNE_MM), y, LARGEUR_COLONNE_MM, hauteur),
  ];
}

/** D1/D2 : demi-page droite de la page 1, sous l'en-tête fixe. */
export function construireZoneD(): [Zone, Zone] {
  const y = Y0_MM;
  const hauteur = HAUTEUR_UTILE_MM - HAUTEUR_ENTETE_MM;
  return [
    zone("D1", 1, colonneX(X_DROITE_MM, 0, LARGEUR_COLONNE_MM), y, LARGEUR_COLONNE_MM, hauteur),
    zone("D2", 1, colonneX(X_DROITE_MM, 1, LARGEUR_COLONNE_MM), y, LARGEUR_COLONNE_MM, hauteur),
  ];
}

/** 4 colonnes strictement identiques occupant toute la page 2. */
export function construireZonesPage2(): Zone[] {
  return [0, 1, 2, 3].map((i) =>
    zone(`C${i + 1}`, 2, colonneX(X0_MM, i, LARGEUR_COLONNE_P2_MM), Y0_MM, LARGEUR_COLONNE_P2_MM, HAUTEUR_UTILE_MM));
}

export interface Grille {
  /** Ordre exact de remplissage des chants : D1 -> D2 -> page2 -> G1 -> G2
   * (G2 exclue si la Prière est active). Jamais l'ordre physique des pages. */
  flowOrder: Zone[];
  /** Toutes les zones par nom, y compris celles hors du flux -- pour que le
   * rendu puisse toujours les localiser (ex: G2 quand la Prière l'occupe). */
  toutes: Record<string, Zone>;
}

/** Si onePageMode est vrai, le flux est restreint à D1/D2 (page 1, moitié
 * droite) et C1/C2 (page 1, moitié gauche) -- jamais de page 2. */
export function construireGrille(priereActive: boolean, onePageMode = false, banniereActive = true): Grille {
  const [d1, d2] = construireZoneD();
  const [g1, g2] = construireZoneG(banniereActive);

  let zonesP2: Zone[];
  let c1: Zone | undefined, c2: Zone | undefined;
  if (onePageMode) {
    const yC = banniereActive ? Y0_MM + HAUTEUR_BANNIERE_MM : Y0_MM;
    const hC = banniereActive ? HAUTEUR_UTILE_MM - HAUTEUR_BANNIERE_MM : HAUTEUR_UTILE_MM;
    c1 = zone("C1", 1, colonneX(X0_MM, 0, LARGEUR_COLONNE_P2_MM), yC, LARGEUR_COLONNE_P2_MM, hC);
    c2 = zone("C2", 1, colonneX(X0_MM, 1, LARGEUR_COLONNE_P2_MM), yC, LARGEUR_COLONNE_P2_MM, hC);
    const c3 = zone("C3", 1, colonneX(X0_MM, 2, LARGEUR_COLONNE_P2_MM), Y0_MM, LARGEUR_COLONNE_P2_MM, HAUTEUR_UTILE_MM);
    const c4 = zone("C4", 1, colonneX(X0_MM, 3, LARGEUR_COLONNE_P2_MM), Y0_MM, LARGEUR_COLONNE_P2_MM, HAUTEUR_UTILE_MM);
    zonesP2 = [c1, c2, c3, c4];
  } else {
    zonesP2 = construireZonesPage2();
  }

  const toutes: Record<string, Zone> = {};
  for (const z of [d1, d2, g1, g2, ...zonesP2]) toutes[z.nom] = z;

  let flow: Zone[];
  if (onePageMode) {
    flow = [d1, d2, c1!];
    if (!priereActive) flow.push(c2!);
  } else {
    flow = [d1, d2, ...zonesP2, g1];
    if (!priereActive) flow.push(g2);
  }
  return { flowOrder: flow, toutes };
}
